megamek/src/megamek/common/MovePath.java

```diff
@@ -311,15 +310,109 @@ protected MovePath addStep(final MoveStep step, boolean compile) {
                 step.setMovementType(EntityMovementType.MOVE_ILLEGAL);
             }
         }
+        
+        // jumping into heavy woods is danger
+        if (game.getOptions().booleanOption(OptionsConstants.ADVGRNDMOV_PSR_JUMP_HEAVY_WOODS)) {
+            Hex hex = game.getBoard().getHex(step.getPosition());
+            if ((hex != null) && isJumping() && step.isEndPos(this)) {
+                PilotingRollData psr = entity.checkLandingInHeavyWoods(step.getMovementType(false), hex);
+                if (psr.getValue() != PilotingRollData.CHECK_FALSE) {
+                    step.setDanger(true);
+                }
+            }
+        }
 
-
-        // check for illegal jumps
         final Coords start = getEntity().getPosition();
         final Coords land = step.getPosition();
+        
+        if (step.getMovementType(false) != EntityMovementType.MOVE_ILLEGAL) {
+            performIllegalCheck(step, start, land);
+        }
+        
+        // If the new step is legal and is a different position than
+        // the previous step, then update the older steps, letting
+        // them know that they are no longer the end of the path.
+        if (step.isLegal(this) && (null != prev) && !land.equals(prev.getPosition())) {
+
+            // Loop through the steps from back to front.
+            // Stop looping when the step says to, or we run out of steps.
+            int index = steps.size() - 2;
+            while ((index >= 0) && getStep(index).setEndPos(false)) {
+                index--;
+            }
+
+        } // End step-is-legal
+        
+        if (shouldMechanicalJumpCauseFallDamage()) {
+            step.setDanger(true);
+        }
+                
+        // If we are using turn modes, go back through the path and mark danger for any turn
+        // that now exceeds turn mode requirement. We want to show danger on the previous step
+        // so the StepSprite will show danger. Hiding the previous step instead would make turning costs
+        // show in the turning hex for units tracking turn mode, unlike other units.
+        if (entity.usesTurnMode() && getMpUsed() > 5) {
+            int turnMode = getMpUsed() / 5;
```

Comment:
Integer division in floating point context

Reply:
I don't know which way it rounds, and I don't want to guess, so I'm going to leave it as is.